### ISSUE TITLE
2D data looks crazy - is it mixing up x0/x1 in some cases?

### ISSUE THREAD

**Issue by hamogu** [sherpa/sherpa#2014]
This issue has a title that's intentionally similar to #840. However, while #840 is about regridding, here, I see things that look like crazy mixups even without regridding.
```Python
import numpy as np
import matplotlib.pyplot as plt

from sherpa.astro.data import DataIMGInt
from sherpa.models import Gauss2D
from sherpa.fit import Fit
from sherpa.stats import Cash

# Use numpy to set up the arrays we want.
# Yes, Sherpa has functions like `dataspace2d`, but I want to be absolutely
# clear what goes in.
x0 = np.random.normal(size=10000, loc=21.6, scale=0.03)
x1 = np.random.normal(size=10000, scale=0.03)
x0range = np.arange(21.4, 21.7, 0.01)
x1range = np.arange(-.25, .250001, 0.01)
hist, x0edges, x1edges = np.histogram2d(x0, x1, bins=(x0range, x1range))
x0lo, x1lo = np.meshgrid(x0edges[:-1], x1edges[:-1])
x0hi, x1hi = np.meshgrid(x0edges[1:], x1edges[1:])

# We made a rectangular dataset and put the signal a bit off-center
# so we can see if the axes are reversed

print(hist.shape)
(30, 50)

image = DataIMGInt("binned_image",
                            x0lo=x0lo.flatten(), x1lo=x1lo.flatten(),
                           x0hi=x0hi.flatten(), x1hi=x1hi.flatten(),
                         y=hist.flatten(), shape=(30, 50))

# MAke a model
gline = Gauss2D("gline")
gline.fwhm = 0.05
gline.fwhm.freeze()
gline.ampl = 115
gline.xpos = 21.6
gline.ypos.frozen = False
gline.ypos = 0
gline.xpos.frozen = False

out = image.get_img(gline)
```
Now, `out` contains two images, the data and the model.
The data looks like this:
![image](https://github.com/sherpa/sherpa/assets/498688/1556c870-abd5-4263-9439-4eab975d7037)
And the model looks like this
![image](https://github.com/sherpa/sherpa/assets/498688/0ed2ca94-4e55-4f30-a0a3-160e8b0e6a0c)

Clearly, the model is borked (technical term I learned from @DougBurke ).

This is clearly a case of x0/x1 mixup. How do I know? If I change the shape that I give the `DataIMGInt` from (30, 50) to (50, 30), then the model is right, but the data is borked.

I've tried fitting, but it's pointless - since data and model are not compatible, you can't get anything to fit. The only way to make that work for now is to use square data, such that `(n1, n2) == (n2, n1)` because `n1==n2`.

Also, but that's just a side-note because it's only for display and to maximize the confusion, see that the notebook plotting calls the second axis x (the one with 50 elements) and the first axis y (the one with 30 elements). instead of calling them `x1` and `x0` (in that order):
<img width="531" alt="Screenshot 2024-04-28 at 4 49 42 PM" src="https://github.com/sherpa/sherpa/assets/498688/d1f4f14b-2605-457f-8916-4d1d28b4201b">

**Comment by hamogu** [comment]
# 1880 and #1789 touch on similar problems, but it's not the same, so I'm opening this as a new issue.

I also note that I experimented with `Data2D` in the code above and I get the impression that the same problem (the model is borked) exists there. It's just not as easy to show, because there is no convenient way to plot Data2D since it could be on an irregular or incomplete grid. (In my example it's not, but it means that plotting to see by eye which axis is which is not straightforward).

**Comment by hamogu** [comment]
Now, we modify the example above to use a square dataset and show that the x0/x1 mixup carries through to fitting and does not just affect the display:
```Python
x0 = np.random.normal(size=10000, loc=21.6, scale=0.03)
x1 = np.random.normal(size=10000, scale=0.03)
x0range = np.arange(21.4, 21.7, 0.01)
x1range = np.arange(-.15, .150001, 0.01)
hist, x0edges, x1edges = np.histogram2d(x0, x1, bins=(x0range, x1range))
x0lo, x1lo = np.meshgrid(x0edges[:-1], x1edges[:-1])
x0hi, x1hi = np.meshgrid(x0edges[1:], x1edges[1:])

image = DataIMGInt("binned_image",
                            x0lo=x0lo.flatten(), x1lo=x1lo.flatten(),
                           x0hi=x0hi.flatten(), x1hi=x1hi.flatten(),
                         y=hist.flatten(), shape=(30, 30))

gline = Gauss2D("gline")
gline.fwhm = 0.05
gline.fwhm.freeze()
gline.ampl = 115
gline.xpos = 21.6
gline.ypos.frozen = False
gline.ypos = 0
gline.xpos.frozen = False

f2d = Fit(data=image, model=gline, stat=Cash())
f2d.fit()
```
gives
```
Parameter | Best-fit value | Approximate error
-- | -- | --
gline.xpos | 21.547 | ± 0.000514376
gline.ypos | 0.0510096 | ± 0.000514103
gline.ampl | 12741.3 | ± 395.732
```

xpos is about in the middle of the range set for `x0` above (from 21.4 to 21.7) *but it should have been at 21.6*. So, it's on the right scale (the number is between 21.4 and 21.7) but it's fitting the wrong image dimension (the one where the the signal is centered). ypos is 0.05, which is just a bit off center on the scale that goes from -0.15 to 0.15. So, ypos is on the right scale, but it's fitting the wrong image dimension (the one where the data is not in the center).

**Comment by hamogu** [comment]
You might be thinking of #840? That's about regridding. It may have the same underlying cause or it may be a separate problem in the regridding code, so I chose to open this as a separate issue.

**Comment by DougBurke** [comment]
Nope, I was definitely thinking of this issue (#840 is related but separate).

**Comment by hamogu** [comment]
This cannot possibly work for all inputs to `DataIMG` because as assume a certain ordering of the x0, x1 values e.g. [here]
(https://github.com/sherpa/sherpa/blob/main/sherpa/astro/data.py#L5726):
```Python
    def get_img(self, yfunc=None):
        # FIXME add support for coords to image class -> DS9
        self._check_shape()
        y_img = self.filter_region(self.get_dep(False))
        y_img = y_img.reshape(*self.shape)
        if yfunc is None:
            return y_img

        m = self.eval_model_to_fit(yfunc)
        if np.iterable(self.mask):
            # if filtered, the calculated model must be padded up
            # to the data size to preserve img shape and WCS coord
            m = pad_bounding_box(m, self.mask)

        return (y_img,
                self.filter_region(m).reshape(*self.shape))
```

The implicitly assumes that the input x0, x1, y are flattened in C order, however, that's not required. Since `DataIMG` inherits from `Data2D` then can be in any order, C, F, or even random. So, fitting should work (because it's using the x0, x1, but the return values reshaped as an image are meaningless of not in C order. 

Since `DataIMG` (and `DataIMGInt`) is meant for images, we could change the API: make x0 a vector of length n, x1 a vector of length m, and y a vector of length n*m. Then, Sherpa itself could use `mgrid` and `flatten` to convert x0 and x1 into n*m vectors and ensure that they are flattened in a predictable way.

This is not a fully worked out API proposal (and also interacts with #1880) at this point, it's just notes for me as I'm investigating this issue.

**Comment by hamogu** [comment]
`sherpa.astro.ui.AstroSession.read_image` works that way (it autogenerates the logical coordinate system from the image dimensions and then runs `mgrid` and `ravel`) which is why all the above always works for images read in with `read_image` - the ordering of the x0, x1 within `DataIMG` is guaranteed to be C-ordered.
